### DISCUSSION TITLE
about App and AppHandle struct

### DISCUSSION THREAD

**Discussion by ggslayer** [tauri-apps/tauri#8526] [Q&A] [answered]
Hi, I'm begining with tauri to build my app, I think rust+js is the best solution what I want on cross platform develop.
But, I'm blocked on the concept of app/apphandle,  which I should save as my SomeApp class's member,  after searching docs, they have similar method, and I don't understand the different bewteen them.

I want save a app instance in my own SomeApp struct, then I can do everything(for example: window create/destory, shortcut setting/ menu setting/ about info create/ ... ) base on it.

I understand plugin is inited on builder and not on App,  because plugin is global concept behind the App.  
But, I don't understand why command handlers is built by builder too, not on the App object?

In the builder doc, there is a [invoke_system](https://docs.rs/tauri/2.0.0-alpha.20/tauri/struct.Builder.html#method.invoke_system) method, which I don't understand it's function and usage...

Is there some method to create app with no builder?     Is there some detail doc to help me create tauri app with lowerest api which can help me got the concept relationships on step by step code to create app's everything like tray/menu/about/event/window/window manage/....

I'm sorry, I have very much confuses with tauri api, because not familar with it and rust,  So, Can you give me some advice?

**Comment by FabianLars** [comment, marked as answer]
You basically always want an `AppHandle`. It's also the only thing you can move into threads for example. You only use `App` if you get it from tauri directly, for example in the Builder's `setup` hook. But then again, once you have to move that into a different context (a thread, or a struct member) you'd want to get an AppHandle via `app.handle()`.

The difference between App and AppHandle is really just that the AppHandle is a handle or a reference to the instance of App (concept wise similar-ish to `&` in rust if that helps). You can have multiple AppHandle instances but only one App == You can have multiple references to your main app but you can only have on actual app.

> I want save a app instance in my own SomeApp struct, then I can do everything(for example: window create/destory, shortcut setting/ menu setting/ about info create/ ... ) base on it.

Use an AppHandle

> I understand plugin is inited on builder and not on App, because plugin is global concept behind the App.

You can register a plugin on the App/AppHandle too https://docs.rs/tauri/latest/tauri/struct.AppHandle.html#method.plugin

>  But, I don't understand why command handlers is built by builder too, not on the App object?

We _probably_ could change that, not sure. But since you can register Plugins at runtime you can register new commands in plugins instead.

> In the builder doc, there is a [invoke_system](https://docs.rs/tauri/2.0.0-alpha.20/tauri/struct.Builder.html#method.invoke_system) method, which I don't understand it's function and usage...

If that's the case you likely don't need it. It was added for this library: https://github.com/tauri-apps/tauri-invoke-http/ which replaces the default invoke system (the IPC, the transport between js and rust), which uses native webview apis, with a standard http server.

> Is there some method to create app with no builder?

No.

> Is there some detail doc to help me create tauri app with lowerest api which can help me got the concept relationships on step by step code to create app's everything like tray/menu/about/event/window/window manage/....

I don't really understand what you're actually asking for. `tauri::Builder` and its methods is the lowest level there is in a tauri app.

**Reply by ggslayer** [reply]
thankyou for your replay,  an aother relative question,  can app run on apphandle?   which  I save to my struct's member by it's clone,  and then, my struct expose a method called run, in which, I want to use my member apphandle to exec run, but I can't find run method in apphandle.

**Reply by FabianLars** [reply]
No, that has to be done on the main Builder or on the App you got from Builder.build. You can't call the run function any other time (like in the setup hook where you have access to `&mut App`.

Now that i say this, you could in theory call Builder.build() and move that App into your struct. It must be kept on the main thread though so you'll hit issues with that approach fairly quickly. Also, `run()` never returns so your code would block on that line.

**Reply by dremnik** [reply]
@FabianLars If you guys get a chance maybe you could add some docs for beginners on working with app handles? Maybe I missed something but wasn't immediately clear that you are supposed to be able to clone an app handle when you need another thread to have access to the app state. Was looking at the state in the commands docs and getting a little confused... just mentioning here in case that gives some perspectives on how beginners might be experiencing the docs (love the framework overall btw)

**Reply by FabianLars** [reply]
Any idea where we should add it? Just in the state guide or somewhere else as well?

**Reply by dremnik** [reply]
I think that would be good, maybe just in the section [Accessing an AppHandle in Commands] it could give a bit of a description about the nature of the AppHandle generally, maybe even a small blurb on accessing app state in a multi-threaded context too?

**Reply by jamesjlee1211-ui** [reply]
“Thanks for the clear explanation on the distinction between App and AppHandle — that was a real sticking point for me. You mention that AppHandle is what you move into threads and store in structs, whereas App is mostly for the builder’s setup. That really helps clarify my mental model.
On a related note for creators doing mobile video editing: if you’re building a desktop companion or tool that manages motion-graphics content made on-the-go, I’ve found [Alight Motion Premium](https://alightmotionpremium.com/blog/alightmotion-mod-apk-download) incredibly helpful. It supports keyframe animation and vector layers on phone, which you could sync to a desktop app built with Tauri—so you get best of mobile + cross-platform.

**Reply by jamesjlee1211-ui** [reply]
it’s really helpful how the distinction between App and AppHandle is explained. I’m working on a mobile-first workflow where editing and creative tools matter a lot, so frameworks like Tauri that bridge Rust + JS are exciting. I’ve been exploring optimized mobile creative tools and modded workflows at [Hypic apk](https://hypicmodapks.org/) to support those kinds of apps. Cheers to the community for the guidance here!
